src/Logic.hs

```diff
@@ -91,150 +88,122 @@ import qualified Time
 
 -- | Leave a comment on the given pull request.
-leaveComment :: PullRequestId -> Text -> Action ()
-leaveComment pr body = liftAction $ LeaveComment pr body ()
+leaveComment :: BaseAction :> es => PullRequestId -> Text -> Eff es ()
+leaveComment pr body = send $ LeaveComment pr body
 
 -- | Check if this user is allowed to issue merge commands.
-isReviewer :: Username -> Action Bool
-isReviewer username = liftAction $ IsReviewer username id
+isReviewer :: BaseAction :> es => Username -> Eff es Bool
+isReviewer username = send $ IsReviewer username
 
-getPullRequest :: PullRequestId -> Action (Maybe GithubApi.PullRequest)
-getPullRequest pr = liftAction $ GetPullRequest pr id
+getPullRequest :: BaseAction :> es => PullRequestId -> Eff es (Maybe GithubApi.PullRequest)
+getPullRequest pr = send $ GetPullRequest pr
 
-getOpenPullRequests :: Action (Maybe IntSet)
-getOpenPullRequests = liftAction $ GetOpenPullRequests id
+getOpenPullRequests :: BaseAction :> es => Eff es (Maybe IntSet)
+getOpenPullRequests = send GetOpenPullRequests
 
-getLatestVersion :: Sha -> Action (Either TagName Integer)
-getLatestVersion sha = liftAction $ GetLatestVersion sha id
+getLatestVersion :: BaseAction :> es => Sha -> Eff es (Either TagName Integer)
+getLatestVersion sha = send $ GetLatestVersion sha
 
-getChangelog :: TagName -> Sha -> Action (Maybe Text)
-getChangelog prevTag curHead = liftAction $ GetChangelog prevTag curHead id
+getChangelog :: BaseAction :> es => TagName -> Sha -> Eff es (Maybe Text)
+getChangelog prevTag curHead = send $ GetChangelog prevTag curHead
 
-getDateTime :: Action UTCTime
-getDateTime = liftEnvironment $ GetDateTime id
+getDateTime :: RetrieveEnvironment :> es => Eff es UTCTime
+getDateTime = send GetDateTime
 
-getBaseBranch :: Action BaseBranch
-getBaseBranch = liftEnvironment $ GetBaseBranch id
+getBaseBranch :: RetrieveEnvironment :> es => Eff es BaseBranch
+getBaseBranch = send GetBaseBranch
 
-getProjectConfig :: Action ProjectConfiguration
-getProjectConfig = liftEnvironment $ GetProjectConfig id
+getProjectConfig :: RetrieveEnvironment :> es => Eff es ProjectConfiguration
+getProjectConfig = send GetProjectConfig
 
-registerMergedPR :: Action ()
-registerMergedPR = liftAction $ IncreaseMergeMetric ()
+registerMergedPR :: BaseAction :> es => Eff es ()
+registerMergedPR = send IncreaseMergeMetric
 
-triggerTrainSizeUpdate :: ProjectState -> Action ()
+triggerTrainSizeUpdate :: BaseAction :> es => ProjectState -> Eff es ()
 triggerTrainSizeUpdate projectState = do
   let n = IntMap.size $ IntMap.filter Pr.isInProgress (Pr.pullRequests projectState)
-  liftAction $ UpdateTrainSizeMetric n ()
+  send $ UpdateTrainSizeMetric n
 
 -- | Interpreter that translates high-level actions into more low-level ones.
-runBaseAction :: ProjectConfiguration -> BaseActionFree a -> Operation a
+runBaseAction
+  :: (MetricsOperation :> es, GitOperation :> es, GithubOperation :> es, TimeOperation :> es)
```

Comment:
Small downside then is that with the concept of `Action` gone, you have to mention the specific individual effects you use in every function.

I could have chosen to type alias this, but I think I prefer it this way, since it shows at one glance what kind of things this function does.

Reply:
I think that if it's to be extended, a type alias would do better here. Depends on how much we rely on the alias, etc.